client/lib/cgutil/cpuset_manager_v2.go

```diff
@@ -0,0 +1,332 @@
+	for id := range c.isolating {
+		if strings.HasPrefix(string(id), allocID) {
+			delete(c.isolating, id)
+		}
+	}
+
+	// recompute available sharable cpu cores
+	c.recalculate()
+
+	// now write out the entire cgroups space
+	c.reconcile()
+
+	// now remove any tasks no longer running
+	c.cleanup()
+}
+
+func (c *cpusetManagerV2) CgroupPathFor(allocID, task string) CgroupPathGetter {
+	// The CgroupPathFor implementation must block until cgroup for allocID.task
+	// exists [and can accept a PID].
+	return func(ctx context.Context) (string, error) {
+		ticks, cancel := helper.NewSafeTimer(100 * time.Millisecond)
+		defer cancel()
+
+		for {
+			path := c.pathOf(makeID(allocID, task))
+			mgr, err := fs2.NewManager(nil, path, rootless)
+			if err != nil {
+				return "", err
+			}
+
+			if mgr.Exists() {
+				return path, nil
+			}
+
+			select {
+			case <-ctx.Done():
+				return "", ctx.Err()
+			case <-ticks.C:
+				continue
+			}
+		}
+	}
+}
+
+// recalculate the number of cores sharable by non-isolating tasks (and isolating tasks)
+//
+// must be called while holding c.lock
+func (c *cpusetManagerV2) recalculate() {
+	remaining := c.initial.Copy()
+	for _, set := range c.isolating {
+		remaining = remaining.Difference(set)
+	}
+	c.pool = remaining
+}
+
+// reconcile will actually write the cpuset values for all tracked tasks.
+//
+// must be called while holding c.lock
+func (c *cpusetManagerV2) reconcile() {
+	for id := range c.sharing {
+		c.write(id, c.pool)
+	}
+
+	for id, set := range c.isolating {
+		c.write(id, c.pool.Union(set))
+	}
+}
+
+// cleanup will remove any cgroups for allocations no longer being tracked
+//
+// must be called while holding c.lock
+func (c *cpusetManagerV2) cleanup() {
+	// create a map to lookup ids we know about
+	size := len(c.sharing) + len(c.isolating)
+	ids := make(map[identity]nothing, size)
+	for id := range c.sharing {
+		ids[id] = null
+	}
+	for id := range c.isolating {
+		ids[id] = null
+	}
+
+	if err := filepath.WalkDir(c.parentAbs, func(path string, entry os.DirEntry, err error) error {
+		// a cgroup is a directory
+		if !entry.IsDir() {
+			return nil
+		}
+
+		dir := filepath.Dir(path)
+		base := filepath.Base(path)
+
+		// only manage scopes directly under nomad.slice
+		if dir != c.parentAbs || !strings.HasSuffix(base, ".scope") {
+			return nil
+		}
+
+		// only remove the scope if we do not track it
+		id := identity(strings.TrimSuffix(base, ".scope"))
+		_, exists := ids[id]
+		if !exists {
+			c.remove(path)
+		}
+
+		return nil
+	}); err != nil {
+		c.logger.Error("failed to cleanup cgroup", "err", err)
+	}
+}
+
+//pathOf returns the absolute path to a task with identity id.
+func (c *cpusetManagerV2) pathOf(id identity) string {
+	return filepath.Join(c.parentAbs, makeScope(id))
+}
+
+// remove does the actual fs delete of the cgroup
+//
+// We avoid removing a cgroup if it still contains a PID, as the cpuset manager
+// may be initially empty on a Nomad client restart.
```

Comment:
I don't think we need to do it for this PR, but it might be nice if we can eventually figure out a way to clean up stray cgroup entries. The allocdir hook is the only hook that comes before the cgroup hook, but if its postrun hook throws an error I think we'll never clean up the cgroup entry for that PID?

Reply:
This is a good point; it's a similar problem for other resources like networks too, right?

Reply:
Yeah https://github.com/hashicorp/nomad/issues/11096 and https://github.com/hashicorp/nomad/issues/6385.